crates/telemetry/src/detector.rs

```diff
@@ -0,0 +1,47 @@
+use std::{env, time::Duration};
+
+use opentelemetry::{Key, KeyValue, Value};
+use opentelemetry_sdk::{
+    resource::{EnvResourceDetector, ResourceDetector},
+    Resource,
+};
+
+const OTEL_SERVICE_NAME: &str = "OTEL_SERVICE_NAME";
+
+/// Custom resource detector for Spin relevant attributes service.name and service.version.
+///
+/// To set service.name this detector will first try `OTEL_SERVICE_NAME` env. If it's not available,
+/// then it will check the `OTEL_RESOURCE_ATTRIBUTES` env and see if it contains `service.name`
+/// resource. If it's also not available, it will use `spin`.
+///
+/// To set service.version, it will use the spin_version passed in new.
+#[derive(Debug)]
+pub struct SpinResourceDetector {
+    spin_version: String,
+}
+
+impl SpinResourceDetector {
+    /// Create a new instance of SpinResourceDetector.
+    pub fn new(spin_version: String) -> Self {
+        SpinResourceDetector { spin_version }
+    }
+}
+
+impl ResourceDetector for SpinResourceDetector {
+    fn detect(&self, _timeout: Duration) -> Resource {
+        let service_name = env::var(OTEL_SERVICE_NAME)
+            .ok()
+            .filter(|s| !s.is_empty())
+            .map(Value::from)
+            .or_else(|| {
+                EnvResourceDetector::new()
+                    .detect(Duration::from_secs(0))
+                    .get(Key::new("service.name"))
+            })
+            .unwrap_or_else(|| "spin".into());
+        Resource::new(vec![
+            KeyValue::new("service.name", service_name),
+            KeyValue::new("service.version", self.spin_version.clone()),
```

Comment:
It would be pretty weird for someone to set service.name to e.g. `MyApp` and for the version to be Spin's version, i.e. "MyApp version 0.1" would effectively look like "MyApp version 2.4.0" in spans.

Maybe we should be setting [`process.runtime.name` and `.version`](https://opentelemetry.io/docs/specs/semconv/attributes-registry/process/) for Spin instead?

Reply:
Agreed about that being weird. But, the user can also set the version if they're setting a custom name e.g. `OTEL_RESOURCE_ATTRIBUTES=service.version=78 OTEL_SERVICE_NAME=mycustomname`. I don't think it is un-reasonable to expect them to do that if they want a custom name. I also think it is important to provide the Spin version out of the box.

I think we should switch to putting this info in `process.runtime.{name, version}` like you suggest in a future PR when we start supporting automatically setting the `service.name` based off the manifest application name. Then we can also automatically set `service.version` based off the manifest application version.